ffs/filcold/filcold.go

```diff
@@ -268,36 +269,52 @@ func (fc *FilCold) makeDeals(ctx context.Context, c cid.Cid, size uint64, cfgs [
 // If the deal finishes successfully it returns a FilStorage result.
 // If the deal finished with error, it returns a ffs.DealError error
 // result, so it should be considered in error handling.
-func (fc *FilCold) WaitForDeal(ctx context.Context, c cid.Cid, proposal cid.Cid) (ffs.FilStorage, error) {
+func (fc *FilCold) WaitForDeal(ctx context.Context, c cid.Cid, proposal cid.Cid, timeout time.Duration) (ffs.FilStorage, error) {
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 	chDi, err := fc.dm.Watch(ctx, []cid.Cid{proposal})
 	if err != nil {
 		return ffs.FilStorage{}, fmt.Errorf("watching proposals in deals module: %s", err)
 	}
 
-	var activeProposal ffs.FilStorage
-	for di := range chDi {
-		switch di.StateID {
-		case storagemarket.StorageDealActive:
-			activeProposal = ffs.FilStorage{
-				ProposalCid:     di.ProposalCid,
-				PieceCid:        di.PieceCID,
-				Duration:        int64(di.Duration),
-				Miner:           di.Miner,
-				ActivationEpoch: di.ActivationEpoch,
-				StartEpoch:      di.StartEpoch,
-				EpochPrice:      di.PricePerEpoch,
+	var last deals.StorageDealInfo
+Loop:
+	for {
+		select {
+		case <-time.After(timeout):
+			msg := fmt.Sprintf("DealID %d with miner %s tracking timed out after waiting for %.0f hours.", last.DealID, last.Miner, timeout.Hours())
+			fc.l.Log(ctx, msg)
+			return ffs.FilStorage{}, ffs.DealError{ProposalCid: proposal, Message: msg}
```

Comment:
So now in `WaitForDeal` we keep watching for the channel to report new status changes of a deal. But we also put a cap of the amount of time we wait between two status changes.

If for example, last reported status was `DealSealing` (which is usually the one that takes more time to change to a next status), and we didn't receive any status update in `timeout` duration, then we'll consider this deal abandoned/errored/etc.

This timeout is what the flag/env names `--ffsdealfinalitytimeout`: _A deal can't be in a non-final status for more than X hours_.

Reply:
Well, yes. Actually, both things are a bit equivalent in practice since mostly the Sealing state is the one that takes hours. The precise description is the first one, and since only Sealing takes hours then the second is true by practical means.

Reply:
I guess I was seeing that the timeout doesn't get reset each time a new non-final status comes in. So that's why it seems like the second description is more accurate.

Reply:
Every time the select is evaluated in the for look, `time.After(..)` returns a new timer with `timeout` duration.
So it gets reset on every received status change.

Makes sense?

Reply:
Aaaahhh yea, got it. Cool thanks for clarifying.